### DISCUSSION TITLE
Push exceeding the 100mb limit

### DISCUSSION THREAD

**Discussion by iMaisho** [community/community#23284] [New to GitHub]
<p>Hey guys,</p>
<p>Sorry for the “repost”, i know there are been plenty of topics on this particular issue, but as a non-native english speaker, a non-programmer and a newbie github user i really struggled understanding the solutions given to this problem.</p>
<p>I’m working on a school project, and I unfortunately commited a 100+mb file. I then tried to push and got the error</p>
<p>So far I :</p>
<ul>
<li>
<p>Reverted the commit so that I get back the files localy</p>
</li>
<li>
<p>Identified and deleted the heavy file localy</p>
</li>
<li>
<p>Commited the delete</p>
</li>
</ul>
<p>But the error is still the same… Help ?</p>
<p>If I ever need to use Git or Git Bash or whatever, please be very clear on your instructions and do not assume I know anything about these… Because I don’t ahah</p>
<p>Thanks a lot </p>

**Comment by lee-dohm** [comment, marked as answer]
<p>If you used the <code>git revert</code> command, I can see where you ran into this problem. This is because git is designed very, very well to prevent people from losing information in their repositories. In almost all cases, a git repository  <strong>only grows</strong> , never shrinks. So, if I’m right about you using the <code>git revert</code> command, here’s what happened:</p>
<ol>
<li>You commit a file larger than 100MB</li>
<li>You attempt to push and get the error</li>
<li>You revert the commit with <code>git revert</code> which  <strong>creates a new commit</strong> where the large file doesn’t exist but leaves the large file in the repository’s history of commits</li>
<li>You attempt to push and get the error again</li>
</ol>
<p>There are also other ways that you could do the above that don’t necessarily require the use of <code>git revert</code>. As for how to solve this problem, let me try to show you in pictures.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/0/0969e3a081d652c93ff70293203601ec33ae5bb0.png" data-download-href="/uploads/short-url/1lha9nqTJV7NNNJLpO5gLiQU3Ys.png?dl=1" title="remove-bad-commit.png"><img src="https://user-images.githubusercontent.com/1038121/181076033-7aabc808-caaa-4a10-900b-611dfa3076c3.png" alt="remove-bad-commit.png" data-base62-sha1="1lha9nqTJV7NNNJLpO5gLiQU3Ys" width="690" height="500" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/0/0969e3a081d652c93ff70293203601ec33ae5bb0_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">remove-bad-commit.png</span><span class="informations">842×2420 83.6 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>So the first set of boxes is the state of things just after you make the commit with the large file. From there, you can execute <code>git reset --soft HEAD^</code>. This instructs git to move the <code>HEAD</code> and <code>master</code> “refs” (assuming <code>master</code> is the name of the branch you have checked out) backwards one commit. <code>git reset</code> moves the <code>HEAD</code> and current branch “ref” or pointer to a new location. The location that I gave in <code>git reset --soft HEAD^</code> is <code>HEAD^</code> (which I pronounce “head hat” because it makes it easy for me to remember). The “hat” means “one commit before”. So <code>HEAD^</code> means “one commit before wherever <code>HEAD</code> is currently pointing”. (If you want to do two commits before, you could specify <code>HEAD^^</code>, etc.)</p>
<p>Now, notice that the “bad commit” still exists in the second set of boxes. This is because git tries really, really, really hard to never forget anything. But, when you push, git only transmits the objects that are “reachable” by the branch you’re pushing. So, even though the bad commit exists on your local machine, if you pushed at this stage, you wouldn’t get the error because git wouldn’t transmit the commit that contains the file that is too large.</p>
<p>So, to get to the third set of boxes, you delete the too large file and then commit everything but it. This creates the new commit on the leftward fork in the image. So when you run <code>git push</code> it shouldn’t try to transmit the bad commit or the too large file.</p>
<p>I hope that helps! Let me know if you have any other questions.</p>

**Comment by cryptoAlgorithm** [comment]
<p>I would recommend that you check out GitHub desktop, a GUI app for committing repositories. It works the same way as git, but has a GUI interface and would stop you from committing to a repository if a file is too large to prevent data loss. This is an official tool written by GitHub and can be downloaded from their webpage.</p>
